README.md
<p align="center"> 
    <img src="https://pbs.twimg.com/profile_images/962741924233994246/1Od1scqZ_400x400.jpg" width="200">
</p>

# Tina
 
 [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com) [![Build Status](https://travis-ci.org/salihciftci/tina.svg?branch=master)](https://travis-ci.org/salihciftci/tina) [![Go Report Card](https://goreportcard.com/badge/github.com/salihciftci/tina)](https://goreportcard.com/report/github.com/salihciftci/tina)

Tina is a Twitter bot who mentions people about their life based on their tweets.

Tina randomly sends mention tweets to random people like "Hello" or something related with their mood that can be examined by Tina.

Basically, Tina's goal is to make people smile. Tina's development activities still going on and it's already alive and accessible.

You can check Tina's [tweets](http://www.twitter.com/TheTinaAI).

## How she works

Basically, Tina searches people's tweet with her specific keywords. After finding people, Tina sends mention tweet to them with her own tweets and tries to make them feel better.

Some of these keywords;

```
"happy",
"terrible",
"stressed",
...
```

A few of these tweets;

```
"You have a great sense of style.",
"I tell everyone how amazing you are.",
"You're an excellent!",
...
```

## Roadmap

- Better search algorithm
- Analyze people's tweet
- More Social Network
- Live without Social Network (:question:)

## Contributing

Please read [CONTRIBUTING.md](https://github.com/salihciftci/tina/blob/master/CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Authors

* **Salih Çiftçi** - [salihciftci](https://github.com/salihciftci)

## License

This project is licensed under the MIT License - see the [LICENSE.md](https://github.com/salihciftci/tina/blob/master/LICENSE) file for details 


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [0.1.2]
### Added
- Tweets

## [0.1.1]
### Removed
- Some  Mention Tweets.
- Some Tweets

## [0.1.0]
### Added
- Initial Commit



tests/tina.py
import requests
import sys
import json
import time
import random
import os.path
from requests_oauthlib import OAuth1
from urllib.parse import quote_plus as encode


def findPeople():
    # Selecting random keyword
    keywords = db["keywords"]
    keyword = keywords[random.randint(0, len(keywords) - 1)]
    print("%s - Tina: Searching people keyword is %s now." %
          (time.ctime(), keyword))

    # Searching people
    sUrl = "https://api.twitter.com/1.1/search/tweets.json?q=%s&vertical=news"
    searchUrl = sUrl % keyword
    r = requests.get(searchUrl, auth=auth)

    if r.status_code == 200:
        randomPersonJson = json.loads(r.text)
    else:
        print("%s - Tina: Unexpected Error, Error Code: %s" %
              (time.ctime(), r.status_code))
        print("%s - Tina: Trying to find people.." % time.ctime())
        findPeople()

    global randomPerson
    randomPerson = []
    while len(randomPerson) < 15:
        try:
            randomPerson.append(randomPersonJson["statuses"][len(
                randomPerson)]["user"]["screen_name"])
        except Exception as e:
            print("%s - Tina: Error: %s" % (time.ctime(), e))
            print("%s - Tina: Trying to find people.." % time.ctime())
            findPeople()


mentionCount = 0


def tweet(status):
    tweetUrl = "https://api.twitter.com/1.1/statuses/update.json?status=%s"

    if status == "tweet":
        randomTweet = tweets[random.randint(0, len(tweets) - 1)]
        r = requests.post(tweetUrl % encode(randomTweet), auth=auth)

        if r.status_code == 200:
            print("%s - Tina: tweeted." % time.ctime())
        elif r.status_code == 403:
            print("%s - Tina: Tweet duplicated." % time.ctime())
            print("%s - Tina: Trying to tweet again.." % time.ctime())
            tweet("tweet")
        else:
            print("%s - Tina: Unexpected Error, Error Code: %s" %
                  (time.ctime(), r.status_code))

    elif status == "mention":
        global mentionCount
        randomMentionTweet = mentionTweets[random.randint(
            0, len(mentionTweets) - 1)]
        randomPersonToTweet = randomPerson[mentionCount]

        r = requests.post(tweetUrl % "%40" + randomPersonToTweet +
                          "%20" + encode(randomMentionTweet), auth=auth)
        if r.status_code == 200:
            print("%s - Tina: Mentioned to %s" %
                  (time.ctime(), randomPersonToTweet))
        elif r.status_code == 403:
            print("%s - Tina: Mention tweet duplicated." % time.ctime())
        else:
            print("%s - Tina: Unexpected Error, Error Code: %s" %
                  (time.ctime(), r.status_code))

        if mentionCount == 14:
            findPeople()
            mentionCount = 0
        else:
            mentionCount += 1


if __name__ == "__main__":
    scriptPath = os.path.dirname(__file__)
    with open(os.path.join(scriptPath, "auth.json")) as file:
        authJson = json.load(file)

    authUrl = "https://api.twitter.com/1.1/account/verify_credentials.json"
    auth = OAuth1(authJson["API_KEY"], authJson["API_SECRET"],
                  authJson["OAUTH_TOKEN"], authJson["OAUTH_TOKEN_SECRET"])

    # Trying to Auth to Twitter
    r = requests.get(authUrl, auth=auth)
    if not r.status_code == 200:
        print("%s - Tina: Auth unsuccessful!" % time.ctime())
        sys.exit()
    else:
        print("%s - Tina: Auth successful!" % time.ctime())

    # Reading db.json for tweets
    with open(os.path.join(scriptPath, "db.json")) as file:
        db = json.load(file)

    tweets = db["tweets"]
    mentionTweets = db["mentionTweets"]

    # Finding People
    findPeople()

    while True:
        for x in range(0, 16):
            tweet("mention")
            time.sleep(901)
        tweet("tweet")
        time.sleep(901)


tina.go
package main

import (
	"encoding/json"
	"io/ioutil"
	"log"
	"math/rand"
	"os"
	"time"
)

// Struct for "db.json"
type Database struct {
	Keywords      []string `json:"keywords"`
	MentionTweets []string `json:"mentionTweets"`
	Tweets        []string `json:"tweets"`
}

var (
	people Tweet
)

const (
	version = "0.1.2"
)

func main() {
	//Checking Auth
	if verify() == 200 {
		log.Println("Auth successful.")
	} else {
		log.Println("Auth unsuccessful.")
		os.Exit(1)
	}

	//Reading Json File
	file, err := ioutil.ReadFile("./db.json")
	if err != nil {
		log.Println(err.Error())
	}

	//Json file to slice
	var db Database
	json.Unmarshal(file, &db)

	// Main loop for Tina
	for {
		//Tweeting to homeline
		if update(db.Tweets[rand.Intn(len(db.Tweets))]) == 200 {
			log.Println("Tweeted to homeline")
		} else {
			log.Println("Couldn't tweet to homeline, skipping.")
		}

		time.Sleep(15 * time.Minute)

		//Finding people
		query := db.Keywords[rand.Intn(len(db.Keywords))]
		json.Unmarshal(search(query), &people)

		//Tweeting to everyone in slice
		for i := 0; i < len(people.Statuses); i++ {
			//Randomly getting tweet from db slice
			tweet := db.MentionTweets[rand.Intn(len(db.MentionTweets))]
			//Mentioning to someone.
			if update("@"+people.Statuses[i].User.ScreenName+" "+tweet) == 200 {
				log.Println("Mention tweeted.")
			} else {
				log.Println("Couldn't tweeted, skipping")
			}

			time.Sleep(15 * time.Minute)
		}
	}
}


tweet.go
package main

import (
	"github.com/dghubble/oauth1"
	"io/ioutil"
	"log"
	"net/url"
	"os"
)

// Setting up Twitter client
var (
	consumerKey    = ""
	consumerSecret = ""
	tokenKey       = ""
	tokenSecret    = ""

	config = oauth1.NewConfig(consumerKey, consumerSecret)
	token  = oauth1.NewToken(tokenKey, tokenSecret)

	client = config.Client(oauth1.NoContext, token)
)

// Authentication client
func verify() int {
	path := "https://api.twitter.com/1.1/account/verify_credentials.json"
	resp, err := client.Get(path)

	if err != nil {
		log.Println(err.Error())
		os.Exit(1)
	}
	defer resp.Body.Close()

	return resp.StatusCode
}

// Tweeting to twitter with client
func update(status string) int {
	path := "https://api.twitter.com/1.1/statuses/update.json"
	resp, err := client.PostForm(path, url.Values{"status": {status}})

	if err != nil {
		log.Println(err.Error())
	}
	defer resp.Body.Close()
	return resp.StatusCode
}

// Searching People with query keyword (from db.json)
func search(query string) []byte {
	path := "https://api.twitter.com/1.1/search/tweets.json?q=" + query
	resp, err := client.Get(path)

	if err != nil {
		log.Println(err.Error())
	}
	defer resp.Body.Close()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		log.Println(err.Error())
	}

	return body
}


tweet.py
import requests, sys, json, time, random
from requests_oauthlib import OAuth1
from urllib.parse import quote_plus

## read db.json file
with open('db.json') as file:
	data = json.load(file)

## auth to twitter api
authUrl = 'https://api.twitter.com/1.1/account/verify_credentials.json'
auth = OAuth1(data['auth']['YOUR_APP_KEY'], data['auth']['YOUR_APP_SECRET'],
 data['auth']['USER_OAUTH_TOKEN'], data['auth']['USER_OAUTH_TOKEN_SECRET'])
r = requests.get(authUrl, auth=auth)

## checking auth
if not r.status_code == 200:
	print("Auth unsuccessful!")
	sys.exit()
else:
	print("Auth successful!")


randomPerson = []
def getPpl():
	global randomPerson
	## select keyword from db.json
	keyword = data["keywords"][random.randint(0,len(data["keywords"])-1)]
	print("Searching keyword is %s" % keyword)

	## trying to find ppl to mention them.
	searchUrl = "https://api.twitter.com/1.1/search/tweets.json?q=%s&vertical=news" % keyword
	r = requests.get(searchUrl, auth=auth)
	searchData = json.loads(r.text) #request to json

	randomPerson = [] #clearing array.

	## getting usernames and putting them in to array.
	count = 0
	while count < 15:
		randomPerson.append(searchData["statuses"][count]["user"]["screen_name"])
		count += 1

## loading tweets
dailyTweets = data["dailyTweets"]
mentionTweets = data["mentionTweets"];

getPpl() ## getting ppl.

## setting up api
url = "https://api.twitter.com/1.1/statuses/update.json?status=%s"

## post daily tweet
def postDailyTweet():
	randomTweet = dailyTweets[random.randint(0,len(dailyTweets)-1)]
	r = requests.post(url % quote_plus(randomTweet), auth=auth)
	if r.status_code == 403:
		print("%s - Tweet Duplicated." % time.ctime())
	else:
		print("%s - Daily Tweet tweeted." % time.ctime())

## post mention tweet
## url encoding (@ = %40)
randomMentionPersonCount = 0
def postMentionTweet():
	global randomMentionPersonCount #local variable to global variable
	randomTweet = mentionTweets[random.randint(0,len(mentionTweets)-1)]
	randomMentionPerson = randomPerson[randomMentionPersonCount]
	r = requests.post(url % "%40" + randomMentionPerson + " " + quote_plus(randomTweet), auth=auth)
	if r.status_code == 403:
		print("%s - Mention Tweet Duplicated." % time.ctime())
	else:
		print("%s - Mention Tweet tweeted to %s." % (time.ctime(), randomMentionPerson))
	randomMentionPersonCount += 1
	if randomMentionPersonCount == 14:
		print("Refreshing people..")
		getPpl()
		randomMentionPersonCount = 0


## action time..
limit = 0
while limit < 1:
	postDailyTweet()
	print("Waiting 15 min for next tweet.")
	time.sleep(901) #wait 15 min for next tweet.
	postMentionTweet()
	print("Waiting 15 min for next tweet.")
	time.sleep(901) #wait 15 min for next tweet.
